include/revng/Model/NamingHelper.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ADT/RecursiveCoroutine.h"
+#include "revng/Model/ArrayType.h"
+#include "revng/Model/CABIFunctionDefinition.h"
+#include "revng/Model/EnumDefinition.h"
+#include "revng/Model/Identifier.h"
+#include "revng/Model/RawFunctionDefinition.h"
+#include "revng/Model/StructDefinition.h"
+#include "revng/Model/UnionDefinition.h"
+#include "revng/Support/CommonOptions.h"
+
+namespace model {
+
+namespace detail {
+
+struct TransparentStringHash {
+  using is_transparent = void;
+
+  std::size_t operator()(const char *Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(const std::string &Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(std::string_view Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(llvm::StringRef Value) const {
+    return std::hash<std::string_view>{}(Value);
+  }
+  std::size_t operator()(const model::Identifier &Value) const {
+    return std::hash<std::string_view>{}(Value.str());
+  }
+};
+
+} // namespace detail
+
+class VerifyHelper;
+
+class NamingHelper {
+private:
+  using GlobalNameCacheType = std::unordered_map<std::string,
+                                                 std::string,
+                                                 detail::TransparentStringHash,
+                                                 std::equal_to<>>;
+
+  GlobalNameCacheType GlobalNameCache = {};
+
+public:
+  const NamingConfiguration &Configuration;
+
+private:
+  NamingHelper(const NamingConfiguration &Configuration) :
+    Configuration(Configuration) {}
+
+  std::optional<std::string_view> isNameForbidden(std::string_view Name);
+  bool populateGlobalNameCache(VerifyHelper &VH, const Binary &Binary);
+
+public:
+  static std::optional<NamingHelper> tryMake(VerifyHelper &VH,
+                                             const Binary &Binary);
+  NamingHelper(const Binary &Binary);
+
+public:
+  [[nodiscard]] bool isGlobalSymbol(std::string_view Name) const {
+    return GlobalNameCache.contains(Name);
+  }
+
+public:
+  Identifier segment(const model::Segment &Segment) const;
+  Identifier function(const model::Function &Function) const;
+  Identifier dynamicFunction(const model::DynamicFunction &Function) const;
+
+  std::string LLVMFunction(const model::Function &Function) const {
+    std::string Result = "local_";
+    if (DebugNames)
+      Result += function(Function).str().str();
+    else
+      Result += Function.Entry().toString();
+
+    return Result;
+  }
+
+public:
+  Identifier type(const model::TypeDefinition &Definition) const;
+
+  Identifier enumEntry(const model::EnumEntry &Entry,
```

Comment:
I kind of disagree 🤔.

I feel like the most important thing should go first, in this case - we're emitting an entry - so _which_ entry we're emitting is the most important bit. All the other arguments are just there for context.

But, if you feel strongly about "a big thing before a small one" rule, we should add it to the code conventions pad or something.

Reply:
I don't have super strong opinions, but usually I like arguments in decreasing order of how generic they are.
Usually "Context" is the first argument, then you have Module, then Function...

Considering the implicit `this` argument, also that is rather generic.
But no biggie.

Reply:
I guess it's the total opposite of how I see things :D

The way I see it, the most generic argument is going to appear a lot more often, as such, no need to draw attention to it as much - it can be towards the end. While the first argument is a good way to indicate "hey, this is a really important bit".

But, yeah, not a strong stance, hence why I suggesting deciding on a single approach in the conventions pad - for consistency sake.

Reply:
I `git grep`'d the LLVM codebase and it seems that it's a bit more common to have the more general thing first.
I will add this to the best pratice.

Reply:
Done!